spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/IcebergVersionFunction.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.functions;
+
+import org.apache.iceberg.IcebergBuild;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.functions.BoundFunction;
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction;
+import org.apache.spark.sql.connector.catalog.functions.UnboundFunction;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/** A function that gets registered to return the current Iceberg version in use from within SQL */
+public class IcebergVersionFunction implements UnboundFunction {
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length > 0) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot bind: %s does not accept arguments", name()));
+    }
+
+    return new IcebergVersionFunctionImpl();
+  }
+
+  @Override
+  public String description() {
+    return name() + " - Returns the runtime Iceberg version";
+  }
+
+  @Override
+  public String name() {
+    return "iceberg_version";
+  }
+
+  /**
+   * Concrete implementation of {@link IcebergVersionFunction} that returns the short version string
+   * when called, e.g. {@code SELECT system.iceberg_version()} will produce results like "0.14.0" or
+   * "0.15.0-SNAPSHOT".
+   */
+  static class IcebergVersionFunctionImpl implements ScalarFunction<UTF8String> {
```

Comment:
Why is this exposed? It should be private.

Reply:
It cannot be private. 

It can be package-private, but making it private makes the invoke method not accessible. I tried several combinations to make it private, but none of them worked.

When invoke is public static and the subclass is private static I get the error message:

interface org.apache.spark.sql.catalyst.expressions.objects.InvokeLike cannot access a member of class org.apache.iceberg.spark.functions.IcebergVersionFunction$IcebergVersionFunctionImpl with modifiers "public static"
java.lang.IllegalAccessException: interface org.apache.spark.sql.catalyst.expressions.objects.InvokeLike cannot access a member of class org.apache.iceberg.spark.functions.IcebergVersionFunction$IcebergVersionFunctionImpl with modifiers "public static"
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:361)
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:591)
	at java.base/java.lang.reflect.Method.invoke(Method.java:558)
Making invoke anything other than public static gives an error about no function invoke defined (indicating it needs to be public). So the best it seems I can do is package-private as a subcl

Reply:
Makes sense! Can you add a comment?

Reply:
Added